cluster/mesos/cluster.go

```diff
@@ -0,0 +1,485 @@
+	}
+
+	if bindingAddress, ok := options.IP("mesos.address", "SWARM_MESOS_ADDRESS"); ok {
+		if bindingAddress == nil {
+			return nil, fmt.Errorf("invalid address %s", bindingAddress)
+		}
+		driverConfig.BindingAddress = bindingAddress
+	}
+
+	if offerTimeout, ok := options.String("mesos.offertimeout", "SWARM_MESOS_OFFER_TIMEOUT"); ok {
+		d, err := time.ParseDuration(offerTimeout)
+		if err != nil {
+			return nil, err
+		}
+		cluster.offerTimeout = d
+	}
+
+	driver, err := mesosscheduler.NewMesosSchedulerDriver(driverConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	cluster.driver = driver
+
+	status, err := driver.Start()
+	if err != nil {
+		log.Debugf("Mesos driver started, status/err %v: %v", status, err)
+		return nil, err
+	}
+	log.Debugf("Mesos driver started, status %v", status)
+
+	return cluster, nil
+}
+
+// RegisterEventHandler registers an event handler.
+func (c *Cluster) RegisterEventHandler(h cluster.EventHandler) error {
+	if c.eventHandler != nil {
+		return errors.New("event handler already set")
+	}
+	c.eventHandler = h
+	return nil
+}
+
+// CreateContainer for container creation in Mesos task
+func (c *Cluster) CreateContainer(config *cluster.ContainerConfig, name string) (*cluster.Container, error) {
+	task, err := newTask(c, config, name)
+	if err != nil {
+		return nil, err
+	}
+
+	go c.pendingTasks.Add(task)
+
+	select {
+	case container := <-task.container:
+		return container, nil
+	case err := <-task.error:
+		return nil, err
+	case <-time.After(5 * time.Second):
+		c.pendingTasks.Remove(task)
+		return nil, strategy.ErrNoResourcesAvailable
+	}
+}
+
+// RemoveContainer to remove containers on mesos cluster
+func (c *Cluster) RemoveContainer(container *cluster.Container, force bool) error {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
+
+	return container.Engine.RemoveContainer(container, force)
+}
+
+// Images returns all the images in the cluster.
+func (c *Cluster) Images() []*cluster.Image {
+	c.RLock()
+	defer c.RUnlock()
+
+	out := []*cluster.Image{}
+	for _, s := range c.slaves {
+		out = append(out, s.engine.Images()...)
+	}
+
+	return out
+}
+
+// Image returns an image with IdOrName in the cluster
+func (c *Cluster) Image(IDOrName string) *cluster.Image {
+	// Abort immediately if the name is empty.
+	if len(IDOrName) == 0 {
+		return nil
+	}
+
+	c.RLock()
+	defer c.RUnlock()
+
+	for _, s := range c.slaves {
+		if image := s.engine.Image(IDOrName); image != nil {
+			return image
+		}
+	}
+
+	return nil
+}
+
+// Containers returns all the containers in the cluster.
+func (c *Cluster) Containers() []*cluster.Container {
+	c.RLock()
+	defer c.RUnlock()
+
+	out := []*cluster.Container{}
+	for _, s := range c.slaves {
+		for _, container := range s.engine.Containers() {
+			if name := container.Config.Labels[cluster.SwarmLabelNamespace+".mesos.name"]; name != "" && container.Names[0] != "/"+name {
+				container.Names = append([]string{"/" + name}, container.Names...)
```

Comment:
This is a bit strange. I see what it does, but container retrieval is probably not the best spot for rewriting names.

For instance, in `Container(IDOrName string)`:
```
+		if container.Id == IDOrName || stringid.TruncateID(container.Id) == IDOrName {
+			return container
+		}
```

Here we forgot to append the name, as well in multiple other places.

Maybe, as a temporary workaround (this is still ugly), you could add a helper function to `Cluster` (`formatContainer`?) which takes care of name cleanup and is called whenever `Cluster` returns a container (`Container()`, `Containers()`, and what about events actually?)